### PR TITLE
Sof 1617: Follow playhead timeline mode

### PR DESCRIPTION
**Pull request by KvelaGorrrrnio**
The PR introduces a timeline component that follows the playhead.
It is only displayed for the active part. The FullSegmentTimelineComponent is used to display the timeline for off-air segments.

Currently the part and piece component used for the FollowPlayheadTimelineComponent are called OffsetableXComponent. The reason it wasn't called FollowPlayheadTimelineXComponent, is that they are very likely to be used with the next task of creating a ScrollableTimelineComponent.

Skærmbillede 2023-09-28 kl  20 06 07

### CODE REVIEW

**Review by LindvedKrvang — APPROVED** [review]
Don't have much to add except it looks really good :)

File: src/app/rundown/components/follow-playhead-timeline/follow-playhead-timeline.component.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+  private getPreviousParts(onAirPartIndex: number): Part[] {
+    const previousParts: Part[] = this.segment.parts
+        .slice(0, onAirPartIndex)
+        .filter(part => this.getDisplayDurationInMs(part) > 0)
+
+    // TODO: Take into account more than the closest previous part and not only if none was found in the first step.
+    // In order to jump in from anywhere and display the correct transition this is needed.
+    if (previousParts.length === 0 && this.segment.parts[onAirPartIndex - 1] && this.isUnplayedPart(this.segment.parts[onAirPartIndex - 1])) {
+      return [this.segment.parts[onAirPartIndex - 1]]
```
Comment (andr9528) [inline src/app/rundown/components/follow-playhead-timeline/follow-playhead-timeline.component.ts:75]: You are extracting the same part multiple times. 
Can't it be changed to only extract it ones?

File: src/app/rundown/components/offsetable-piece/offsetable-piece.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { OffsetablePieceComponent } from './offsetable-piece.component';
+import { Piece } from '../../../core/models/piece'
+import { instance, mock } from '@typestrong/ts-mockito'
+
+describe('PieceComponent', () => {
+  let component: OffsetablePieceComponent;
```
Comment (andr9528) [inline src/app/rundown/components/offsetable-piece/offsetable-piece.component.spec.ts:8]: Mismatch between component being testet and the string given to the describe, unless i am mistaken, for some reason.

Comment (andr9528) [reply src/app/rundown/components/offsetable-piece/offsetable-piece.component.spec.ts:8]: Any reason why `OffsetablePieceComponent.name`, is not being used?
This is not the only test file where that isn't used.

File: src/app/rundown/components/follow-playhead-part/part.component.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import { RundownService } from '../../../core/abstractions/rundown.service'
+import { PartEntityService } from '../../../core/services/models/part-entity.service'
+
+@Component({
+  selector: 'sofie-part',
+  templateUrl: './part.component.html',
+  styleUrls: ['./part.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class PartComponent implements OnChanges {
```
[the remark this answers is not in the document]
Comment (KvelaGorrrrnio) [reply src/app/rundown/components/follow-playhead-part/part.component.ts]: Was due to WIP – is renamed.

File: src/app/rundown/components/follow-playhead-part/part.component.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  public pieceLayers: PieceLayer[]
+
+  @Input()
+  public pixelsPerSecond: number
+
+  public layeredPieces: Piece[][] = []
+  public partWidthInPixels: string = '0px'
+  public duration: number = 4000
+  // TODO: Remove this.
+  public DEFAULT_PART_DURATION_IN_MS = 4000
```
[the remark this answers is not in the document]
Comment (KvelaGorrrrnio) [reply src/app/rundown/components/follow-playhead-part/part.component.ts]: The entire component is removed in #29.

File: src/app/rundown/components/offsetable-piece/offsetable-piece.component.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  @HostBinding('style.width.px')
+  public get widthInPixels(): number {
+    const displayDurationInMs: number = this.getDisplayDurationInMs()
+    return Math.floor(this.pixelsPerSecond * displayDurationInMs / 1000)
+  }
+
+  private getDisplayDurationInMs(): number {
+    // TODO: Use postPlayheadInset instead of magic numbers
```
[the remark this answers is not in the document]
Comment (KvelaGorrrrnio) [reply src/app/rundown/components/offsetable-piece/offsetable-piece.component.ts:46]: I have removed the outdated todo and extracted the magic number on line 69.

File: src/app/rundown/components/offsetable-piece/offsetable-piece.component.spec.ts (hunk 2 above)
Comment (KvelaGorrrrnio) [reply src/app/rundown/components/offsetable-piece/offsetable-piece.component.spec.ts:8]: Yes. This is autogenerated and then the component name has been changed. I'll fix it.

File: src/app/rundown/components/follow-playhead-timeline/follow-playhead-timeline.component.ts (hunk 1 above)
Comment (KvelaGorrrrnio) [reply src/app/rundown/components/follow-playhead-timeline/follow-playhead-timeline.component.ts]: Has been moved to a variable called `partBeforeOnAirPart`.

File: src/app/rundown/components/offsetable-piece/offsetable-piece.component.spec.ts (hunk 2 above)
Comment (andr9528) [reply src/app/rundown/components/offsetable-piece/offsetable-piece.component.spec.ts:8]: Soooo, when's the change comming? :)

File: src/app/rundown/components/timeline-markers/timeline-markers.component.spec.ts [hunk 6]
```diff
@@ -2,7 +2,7 @@ import { TestBed } from '@angular/core/testing'
 
 import { TimelineMarkersComponent } from './timeline-markers.component'
 import { instance, mock } from '@typestrong/ts-mockito'
-import { MillisecondsAsTimePipe } from '../../../shared/pipes/millisecondsAsTime.pipe'
+import { TimestampPipePipe } from '../../../shared/pipes/timestamp.pipe'
```
Comment (andr9528) [inline src/app/rundown/components/timeline-markers/timeline-markers.component.spec.ts:5]: PipePipe? :D

**Review by andr9528 — APPROVED** [review]
As long as the woopsy with the nameing is fixed, it am hapy with it.

File: src/app/rundown/components/timeline-markers/timeline-markers.component.spec.ts (hunk 6 above)
Comment (KvelaGorrrrnio) [reply src/app/rundown/components/timeline-markers/timeline-markers.component.spec.ts:5]: It is fixed in https://github.com/tv2/ng-sofie/pull/29.

**Pull request merged**